File: UnityGame/Assets/Plugins/Libs/Proto/ProtoS.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System;

public partial class ProtoS
{
	public static Dictionary<int, IProtoItem> 		opcodeDict 		= new Dictionary<int, IProtoItem>();
	public static Dictionary<Type, IProtoItem> 		typeDict 		= new Dictionary<Type, IProtoItem>();
	public static Dictionary<Type, IProtoItem> 		aliasTypeDict 	= new Dictionary<Type, IProtoItem>();

	/** 添加 */
	public static void AddItem(IProtoItem item)
	{
		opcodeDict.Add 			(item.opcode				, item);
		typeDict.Add 			(item.protoStructType		, item);
		aliasTypeDict.Add 		(item.protoStructAliasType	, item);
	}

	/** 获取 ProtoItem, 用 opcode */
	public static IProtoItem GetItemByOpcode(int opcode)
	{
		if (!opcodeDict.ContainsKey (opcode)) 
		{
			Loger.LogTagErrorFormat ("SocketNet", "ProtoS 不存在opcode={0}的ProtoItem", opcode);
		}
		return opcodeDict[opcode];
	}

	/** 获取 ProtoItem, 用 Type */
	public static IProtoItem GetItemByType(Type type)
	{
		if (typeDict.ContainsKey (type)) 
		{
			return typeDict[type];
		}


		if (aliasTypeDict.ContainsKey (type)) 
		{
			return aliasTypeDict[type];
		}

		Loger.LogTagErrorFormat ("SocketNet", "ProtoS 不存在type={0}的ProtoItem", type);
		return null;
	}

	/** 获取 ProtoItem, 用 msg */
	public static IProtoItem GetItemByMsg(object msg)
	{
		return GetItemByType(msg.GetType());
	}


}

File: UnityGame/Assets/Game/Lua/gameframe/net/ProtoS.lua
ProtoS = class("ProtoS", {
	-- <opcode, ProtoItem>
	opcodeDict 	= {},

	-- <name, ProtoItem>
	nameDict 	= {},

	-- <Struct, ProtoItem>
	structDict 	= {},
})
local M = ProtoS



-- 添加 ProtoItem
function M:AddItem( protoItem )
	self.opcodeDict	[protoItem.opcode] 				= protoItem
	self.nameDict	[protoItem.protoStructName] 	= protoItem
	self.structDict	[protoItem.protoStruct] 		= protoItem
end


-- 获取 ProtoItem, 用 opcode
function M:GetItemByOpcode( opcode )
	return self.opcodeDict[opcode]
end

-- 获取 ProtoItem, 用 protoStructName
function M:GetItemByName( protoStructName )
	return self.nameDict[opcode]
end

-- 获取 ProtoItem, 用 protoStruct
function M:GetItemByStruct( protoStruct )
	return self.structDict[protoStruct]
end



-- 获取 ProtoItem, 用 msg
function M:GetItemByMsg( msg )
	local name = getmetatable(msg)._descriptor.name
	return self.nameDict[name]
end



-- 获取 opcode, 用 msg
function M:GetOpcodeByMsg( msg )
	local item = self:GetItemByMsg(msg)
	return item.opcode
end

